src/GafferArnold/IECoreArnoldPreview/Renderer.cpp

```diff
@@ -919,6 +936,7 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
 			{
 				m_polyMesh.apply( mesh, node );
 				m_displacement.apply( node );
+				AiNodeSetByte( node, g_autobumpVisibilityArnoldString, m_visibility_autobump );
```

Comment:
Since `m_displacement.apply()` is responsible for setting `g_dispAutoBumpArnoldString` and the other displacement options, I wonder if it should also be taking care of this? That would fix a bug : `hashProceduralGeometry()` needs to account for the autobump visibility, and it already calls `m_displacement.hash()`, so moving the hashing to `m_displacement` would mean it applied in both places.

Reply:
Sorry, I don't think I explained this very well. We have a bug where we are not appending `m_visibilityAutoBump` in `hashProceduralGeometry()`. Because we are not appending it, we would think that two procedurals were the same even though they had different autobump settings, and we would "deduplicate" them by instancing, so one of them would be wrong. To fix that we need to include `m_visibilityAutoBump` in `hashProceduralGeometry()`, one way or another.

What I was trying to suggest but wasn't describing clearly was this :

- Move `m_visibilityAutoBump` to `Displacement::m_autobumpVisibility`, because it is a displacement setting, and only needs setting on nodes which support displacement.
- Update `Displacement::hash()` to include `m_autobumpVisibility`.
- Now `hashGeometryInternal()` and `hashProceduralGeometry()` will both be correct, because they are delegating to `Displacement::hash()` already.

Does that make sense? I can rejig it if you'd prefer, since we're so close now...

Reply:
Oh yeah alright, I haven't conceived it as our first idea was to make it along `m_visibility`, but this makes sense for the reason you mentioned earlier.

Reply:
So I moved the variable to Displacement struct and renamed it to `autoBumpVisibility` to better blend with the other variables. Can you confirm the updateVisibility stays in the `ArnoldAttributes` constructor, becoming e.g. `updateVisibility( m_displacement.autoBumpVisibility, g_cameraVisibilityAutoBumpAttributeName, AI_RAY_CAMERA, attributes );` ? (I would have liked moving it, but as its private, I won't modify everything)
I also moved the var init, the byte setting and the hash inside of the Displacement.